### ISSUE TITLE
PyInstaller-made app executable size balloons 10x when going from poetry 1.1 to poetry 1.3 or poetry 1.5 

### ISSUE THREAD

**Issue by hintron** [python-poetry/poetry#8323]
- **Poetry version**: 1.5.1 (but reproducable on Poetry 1.3 as well)
- **Python version**: 3.10.12
- **OS version and name**: Linux - SUSE 12.5
- **pyproject.toml**: https://gist.github.com/hintron/f9b5f04a0052602caa9bc8db052d6500

- [x] I am on the [latest](https://github.com/python-poetry/poetry/releases/latest) stable Poetry version, installed using a recommended method.
- [x] I have searched the [issues](https://github.com/python-poetry/poetry/issues) of this repo and believe that this is not a duplicate.
- [x] I have consulted the [FAQ](https://python-poetry.org/docs/faq/) and [blog](https://python-poetry.org/blog/) for any relevant entries or release notes.
- [x] If an exception occurs when executing a command, I executed it again in debug mode (`-vvv` option) and have included the output below.

## Issue
We use Conda and the conda forge channel to set up poetry, along with other non-python tools. We also use PyInstaller to package our Python tool into a self-contained executable, and it seems to work well enough for our uses.

The problem is that when I upgraded poetry from 1.1 to 1.3 in January 2023, the executable binary size created by PyInstaller ballooned 10x, from around 50 MB to 482 MB. This is obviously unacceptable, and the only change was upgrading poetry. So we decided to leave poetry at 1.1 for the near term.

Today, I upgraded poetry from 1.1 to 1.5 (with PyInstaller v5.13.0), hoping that perhaps the issue resolved itself over the course of the last two releases. It did not. I hit the same 10x size increase in the executable size.

My first attempt at hoping the problem goes away over time didn't work, so now I'm looking for ideas :) It seems like newer versions of poetry download more wheels - could that be it? I also noticed that PyInstaller seems to pick up way more files after poetry 1.3+, including a bunch of static library .a files, when before it was just .so files (part of the wheels?).

So I'm really not quite sure how to continue debugging this, or how to get my app back down to 50 MB in size with poetry 1.5. But I don't love the idea of staying at poetry 1.1, as things may start to bit rot over time.

**Comment by dimbleby** [comment]
I expect you'll need to understand better what pyinstaller is doing, exactly what files it is including that you think it ought not to include, and what - if anything - that has to do with poetry.  

Pretty much all of that starts with understanding pyinstaller, not poetry.

**Comment by Secrus** [comment]
> I expect you'll need to understand better what pyinstaller is doing, exactly what files it is including that you think it ought not to include, and what - if anything - that has to do with poetry.
> 
> Pretty much all of that starts with understanding pyinstaller, not poetry.

I second that. When (and if) you find Poetry to be at fault, we would be happy to discuss what should be fixed/changed. Until then, I am going to close this issue, since it's unactionable on our end.

**Comment by hintron** [comment]
Ok, after talking with the PyInstaller guys, it seems that poetry 1.1 is installing numpy from wheels, while poetry 1.3/1.5 is installing numpy from Anaconda packages (https://github.com/pyinstaller/pyinstaller/issues/7856#issuecomment-1682350051). And PyInstaller currently doesn't handle symlinks well.

Is there any way in poetry that I can go back to the behavior of using wheels instead of Anaconda packages? Any insight into what/why/how this behavior changed?

**Comment by dimbleby** [comment]
I expect that poetry being able to use conda packages that were already available in the environment would have been considered a bug fix at some point.

if you don't want to use conda-provided wheels - then simply don't use conda?

**Comment by hintron** [comment]
It would be nice if there was a flag in poetry to go back to the pre-"bug fix" behavior of not trying to reuse things already in the conda environment.

We use conda to install poetry as well as other non-python dependencies into our dev environment, so it's not as simple as "don't use conda," unfortunately.

**Comment by hintron** [comment]
> I second that. When (and if) you find Poetry to be at fault, we would be happy to discuss what should be fixed/changed.
@Secrus I feel like I've demonstrated that Poetry is at fault here. I never told it to use the Conda packages available in my dev environment, which balloon my PyInstaller executable 10x in size. I just want it to do whatever it was doing before. Is there anything we can do here?

**Comment by hintron** [comment]
@dimbleby @Secrus I would like to reopen this issue.

What is causing the large size increase is that Poetry is pulling numpy with Intel MKL from the existing conda environment, while before, poetry pulled numpy from PyPI, which uses OpenBLAS instead of Intel MKL.

I just want to go back to using numpy with OpenBLAS instead of Intel MKL.

From https://numpy.org/install/:

> The NumPy wheels on PyPI, which is what pip installs, are built with OpenBLAS...
In the conda defaults channel, NumPy is built against Intel MKL.
The MKL package is a lot larger than OpenBLAS, it’s about 700 MB on disk while OpenBLAS is about 30 MB.

So how do I get NumPy back to being pulled from PyPI instead of the conda defaults channel? Can I tell poetry to explicitly pull NumPy from PyPI?

I tried doing this

```
numpy = { version = "^1.23.4", source = "pypi" }
```

but that did not change anything.

**Comment by dimbleby** [comment]
if you don't want to use the conda-provided numpy, then don't install the conda-provided numpy

**Comment by hintron** [comment]
Great. That is what I want. But I don't request numpy in my conda environment. Here is my `environment.yml`:

```yml
# miniconda development environment
name: xxxxxxxxxxxx
channels:
  - defaults
  - conda-forge

dependencies:
  - conda-forge::poetry=1.3
  - python=3.10
  - pip
  - go-shfmt=3.3.0
  - shellcheck=0.6.0
  - ghp-import=2.0.1
  - gh
  - jq
  - ruby
```

I only request numpy to be installed via poetry in my `pyproject.toml`, and I request it from pypi. So it's unclear to me how conda is installing numpy to begin with:

```
$ conda list numpy
# packages in environment at /nfs/site/disks/aadg.hintonmi.1/code/trace-list-toolkit/.devenv/envs/trace-list-toolkit:
#
# Name                    Version                   Build  Channel
numpy                     1.26.2                   pypi_0    pypi
numpy-base                1.26.0          py310hb5e798b_0  
```

**Comment by hintron** [comment]
Ok, I confirmed that numpy is being installed by conda before I even call `poetry install`.

It turns out that _poetry itself_ is what requires numpy to be installed in my conda environment. Here is the dependency chain (derived from [conda tree][1] -> `conda tree deptree --full`):

`poetry 1.3.2 -> cleo 2.0.1 -> rapidfuzz 2.13.7 -> numpy 1.26.0`

So now I am in a chicken and egg situation. I want conda to install poetry, but in order to do that, conda has to install numpy. So I can't just not install a conda-provided numpy, like you suggest @dimbleby. I am missing something here? What are my options now?

[1]: https://github.com/conda-incubator/conda-tree

**Comment by dimbleby** [comment]
none of this is anything to do with poetry

- poetry has no dependency on numpy, either direct or indirect
- you shouldn't be mixing the poetry environment with your project environment anyway

**Comment by hintron** [comment]
> none of this is anything to do with poetry
> 
> * poetry has no dependency on numpy, either direct or indirect

That's not entirely true.
https://github.com/python-poetry/poetry/blob/6e6507be57ed745f1c4976966c02464d63aafd8a/poetry.lock#L1317
So I'm guessing that whatever conda-forge did to package poetry 1.3.2 made rapidfuzz include numpy, while a normal poetry install doesn't do that. I'm not quite sure how that all works, though, or if I have any hope of changing the conda-forge package to not install numpy.

> * you shouldn't be mixing the poetry environment with your project environment anyway

I don't know what you mean. Can you elaborate? I'm using conda to bootstrap poetry and to install other non-Python tools in my dev environment. Thanks

**Comment by dimbleby** [comment]
(i) it is entirely true (ii) to learn how to install poetry, read the poetry docs

**Comment by hintron** [comment]
Thanks for the tip.

For anyone else who may find this in the future, I was able to solve this issue by isolating poetry from my app's conda dev environment by creating a second conda environment that just installed poetry. Then, I symlinked that poetry binary into my app's dev environment. This made it so the version of numpy that my app used was from pypi by default and not reused from the numpy used for conda-forge's poetry, and now my app's executable created by PyInstaller is back down to 56 MB (compressed).
